src/linker/configuration.py

```diff
@@ -234,3 +220,100 @@ def _get_requests(environment: Dict[Any, Any], key: str) -> Dict[Any, Any]:
                         f"Assigning default value for {key} {default_key}: '{default_value}'"
                     )
         return requests
+
+    def _get_schema(self) -> Optional[PipelineSchema]:
+        """Validates the pipeline against supported schemas.
+
+        NOTE: this acts as the pipeline configurat file's validation method since
+        we can only find a matching schema if the file is valid.
+        """
+        errors = defaultdict(dict)
+        # Check that the pipeline specification contains a single "steps" outer key
+        if not len(self.pipeline) == 1 or not "steps" in self.pipeline:
+            errors[PIPELINE_ERRORS_KEY][
+                "generic"
+            ] = "The pipeline specification should contain a single 'steps' key."
+            exit_with_validation_error(dict(errors))
+
+        # Check that each of the pipeline steps also contains an implementation
+        metadata = load_yaml(IMPLEMENTATION_METADATA_PATH)
```

Comment:
The whole section below this (and several other places) you look at a dict and you do some sort of validation or parsing (getting the schema). I assume you have tried to make helper functions to make your life easier but there seems to be a lot of use cases to handle. If you aren't able to extract out a helper function I think you could at least make the handling of your errors a bit easier.

Reply:
Have a function that updates errors for you. So upfront you could define some object that has the keys you want to look for and the associated error message and then each if/elif/else branch you can just call that function.  I think this would make this a bit easier to read as well.

Reply:
Yeah I agree with the sentiment here. Have we considered whether any common validation packages might work for us here? https://stackoverflow.com/questions/3262569/validating-a-yaml-document-in-python
e.g. json-schema, schema, pydantic...some of these may also be able to set default values.

Reply:
Yeah there's a lot going on here and it seems more and more like now is the time to do some heavy refactoring. It's tricky because (1) we are trying to implement default values but know we should just move to vivarium's ConfigTree, (2) we are trying to validate which has been scattered across multiple methods of different objects, and (3) we are very specifically trying to batch and log the validation errors rather than just raising at the first one. These three issues are all jumbled now.

I agree that something needs to change. I'll take a look again w/ fresh eyes.

I have not myself messed w/ any external validation packages.